### PR TITLE
[rocjitsu][waitcheck] Detect memory hazards in instruction streams

### PR DESCRIPTION
**Pull request by kuhar**
Add a model-only waitcheck analyzer for ordinary memory dependencies in straight-line instruction streams. Report missing waits with producer and consumer locations, retain committed VGPR reads on CDNA, and account for GFX12 source lifetimes and translation ordering. This builds on #11484.

Writes remain dependent on every pending destination. Same-kind VMEM overwrites are exempt only on architectures with ordered VGPR writeback; counter retirement order alone is insufficient on GFX12. Scalar-buffer accesses use decoded data and descriptor widths.

Return explicit incomplete coverage at control transfers, relative-register operands, mode changes, and deferred instruction families, including parity-ordered LDS loads. Recognize PC destinations as well as branch flags, and use the shared program-path termination predicate for abort traps. Invalid inputs return util::FailureOr. A symbol-boundary check keeps the checker archive independent of the default ISA registry and runtime implementations.

Validation:
- Seven regression groups reproduced the review findings before the fixes; all 103 focused tests pass with ASan, UBSan, and expensive checks.
- GCC compiles the changed implementation and tests with warnings as errors. Injecting a default-registry definition now fails the archive guard; the original guard incorrectly accepted it.
- A local semantic backport onto shared/rocjitsu/sanitizers passes 1,169 waitcheck, C API, and ConSan tests.
- On physical gfx1201 hardware with PyTorch 2.11, eager softmax and fresh-cache torch.compile softmax pass with fail-on-hazard enabled. Top-k passes its numerical oracle in reporting mode with identical existing diagnostics before and after; strict mode already fails on those diagnostics in the baseline.
- The gfx1250 simulator demo with the combined hook and maximum ConSan sampling reports both the intentional missing wait and LDS race and completes successfully.
- One local peanut-review round completed; findings addressed or tracked for the CFG follow-up.

### CODE REVIEW

**Review by atgutier — COMMENTED** [review]
[MAJOR] /tmp/pr11771-review.STx2Y7/emulation/rocjitsu/lib/rocjitsu/src/rocjitsu/code/analysis/
  waitcheck/stream.cpp:190 — WAW exemptions suppress required waits on multiple architectures

  Reasoning: The checker suppresses two unsafe write-after-write cases:

  - ordered_waw() treats same-kind VMEM loads as ordered on every architecture. RDNA4/CDNA5 do not
    guarantee in-order VGPR writeback, so two loads targeting the same VGPR can complete in reverse
    order without a diagnostic.

  - On CDNA3/4, lines 247–260 remove any non-memory destination from a pending producer’s dependency
    set. Additionally, line 253 removes old_value_regs before both read and write checks.
    Consequently, a pending load followed by v_mov_b32 to the same VGPR is reported clean even though
    the late load can overwrite the VALU result.

  Evidence: /tmp/pr11771-review.STx2Y7/emulation/rocjitsu/tests/analysis/waitcheck_stream_test.cpp:128
  explicitly expects two RDNA4 loads to the same destination to be clean, while /tmp/pr11771-
  review.STx2Y7/emulation/rocjitsu/tests/analysis/waitcheck_stream_test.cpp:240 expects CDNA memory-
  to-VALU overwrites to be clean. Public LLVM only exempts same-kind VMEM WAW when the subtarget
  advertises ordered VMEM VGPR writes (/home/agutierr/rocm-dev/llvm-project/llvm/lib/Target/AMDGPU/
  SIInsertWaitcnts.cpp:2493); GFX12 omits that property (/home/agutierr/rocm-dev/llvm-project/llvm/
  lib/Target/AMDGPU/AMDGPU.td:1713).

  Suggested fix: Use event.regs - event.old_value_regs only for reads; write dependencies must start
  from the full destination set. Suppress same-kind VMEM WAW only behind an architecture capability
  equivalent to ordered VMEM VGPR writeback. Add RDNA4/CDNA5 load→load and CDNA3/4 load→VALU
  regressions.

  Confidence: high; compiler behavior and target properties were checked, but not validated on
  hardware.

[MAJOR] /tmp/pr11771-review.STx2Y7/emulation/rocjitsu/lib/rocjitsu/src/rocjitsu/code/analysis/
  waitcheck/stream.cpp:127 — Unflagged control transfers are analyzed as fallthrough

  Reasoning: is_control_flow() recognizes only four instruction flags. Several decoded PC-writing
  instructions—including s_rfe_b64, s_rfe_restore_b64, s_cbranch_join, s_cbranch_g_fork, and CDNA5
  s_rfe_i64—do not carry those flags. Because they have operands, the fallback at lines 158–161 does
  not reject them either. The analyzer therefore processes bytes after a transfer as fallthrough; a
  later wait can retire pending state and yield a false complete, clean report.

  The loop also stops only on PROGRAM_TERMINATOR at line 416, bypassing the existing
  is_program_path_terminator() handling for non-returning s_trap 2.

  Evidence: Generated definitions expose PC destinations without control-flow flags, for example
  generated/cdna3/sop1.cpp:850–860 and generated/cdna3/sop2.cpp:1432–1513. The shared terminal
  predicate is defined in code/analysis/control_flow.h:15–27.

  Suggested fix: Correct the authoritative instruction metadata so every PC-changing operation is
  classified, use that shared predicate for the completeness boundary, and use
  is_program_path_terminator() when deciding where the input stream ends. Add pending-load → s_rfe_* →
  wait/use and s_trap 2 → trailing bytes regressions.

  Confidence: high; the false-fallthrough path follows directly from the decoded operands, zero flags,
  and loop conditions.

  ## MINOR

  [MINOR] /tmp/pr11771-review.STx2Y7/emulation/rocjitsu/lib/rocjitsu/src/rocjitsu/code/analysis/
  waitcheck/stream.cpp:111 — Scalar-buffer store data is widened to four SGPRs

  Reasoning: Source operand zero of every s_buffer_* instruction is forcibly widened to four SGPRs.
  For s_buffer_store_dword, however, source zero is the one-SGPR store value; the four-SGPR descriptor
  is source one. A pending load into s5 followed by a store of s4 is therefore treated as reading
  s[4:7] and produces a false missing-wait diagnostic.

  Evidence: generated/cdna4/smem.cpp:562–574 defines source zero as 32-bit sdata and source one as
  128-bit sbase; the x2 form similarly defines a two-SGPR sdata at lines 590–601.

  Suggested fix: Preserve the decoded operand width. If any load encoding still needs correction,
  restrict it to the operand and instruction forms whose generated metadata is actually deficient.
  Test non-overlapping sdata, true sdata overlap, and descriptor overlap separately.

  Confidence: high.

[MINOR] /tmp/pr11771-review.STx2Y7/emulation/rocjitsu/lib/rocjitsu/src/rocjitsu/code/analysis/
  waitcheck/stream.cpp:143 — v_swaprel_b32 bypasses the relative-register completeness gate

  Reasoning: The gate recognizes only v_movrel* and s_movrel*. v_swaprel_b32 also derives effective
  VGPR operands through M0, but the checker analyzes only its encoded registers and still marks the
  report complete. Pending results in the effective register can therefore be missed or attributed to
  the wrong register.

  Evidence: code/analysis/liveness.cpp:90–100 already classifies both v_movrel and v_swaprel as
  indirect VGPR access. generated/rdna4/vop1.cpp:3775–3788 shows v_swaprel_b32 using M0 with read/
  write VGPR operands.

  Suggested fix: Reuse a shared indirect-register predicate in the stream completeness gate and add
  v_swaprel_b32 to RelativeRegisterOperandsAreIncomplete.

  Confidence: high.

  [MINOR] /tmp/pr11771-review.STx2Y7/emulation/rocjitsu/tests/check_waitcheck_stream_symbols.cmake:27
  — Registry ownership check searches for a nonexistent symbol

  Reasoning: The archive-boundary test looks for rocjitsu::isa_target_registry(), but the actual
  default-composition entry point is rocjitsu::default_isa_target_registry(). The check therefore
  passes even if the waitcheck archive accidentally acquires the default registry it is intended to
  forbid.

  Evidence: The real declaration is target_registry.h:139, its definition is
  target_registry_composition.cpp:18, and no isa_target_registry() function exists. The present
  archive composition appears correct; the defect is in the new regression guard.

  Suggested fix: Search for the exact demangled rocjitsu::default_isa_target_registry() symbol,
  optionally also forbidding the enum-based Decoder::create(rj_code_arch_t) composition entry point.

  Confidence: high.

File: emulation/rocjitsu/lib/rocjitsu/src/rocjitsu/code/analysis/waitcheck/target.h [hunk 1]
```diff
@@ -77,6 +77,12 @@ enum class WaitcntModel { LegacyNoVscnt, LegacyVscnt, SplitGfx12 };
   return arch == ROCJITSU_CODE_ARCH_RDNA4 || arch == ROCJITSU_CODE_ARCH_CDNA5;
 }
 
+// Counter retirement order does not imply ordered VGPR writeback on GFX12.
+[[nodiscard]] inline bool has_ordered_vmem_writeback(rj_code_arch_t arch) {
+  return arch == ROCJITSU_CODE_ARCH_CDNA3 || arch == ROCJITSU_CODE_ARCH_CDNA4 ||
```
Comment (accauble) [inline emulation/rocjitsu/lib/rocjitsu/src/rocjitsu/code/analysis/waitcheck/target.h]: Nit: This seems like an important thing to get right when adding a new architecture. Might be safer to list them out.

**Review by atgutier — COMMENTED** [review]
[MAJOR] emulation/rocjitsu/lib/rocjitsu/src/rocjitsu/code/analysis/waitcheck/stream.cpp:86 — Legal
  scalar-memory destinations can disappear from dependency tracking

  register_effects() collects explicit operands only through to_register_ref() and stores them in
  RegisterSet. This loses valid scalar destinations in two ways:

  - s_load_dword ttmp0, s[0:1], 0 produces a TTMP reference, but RegisterSet::expand() deliberately
    drops all TTMP registers. The resulting event has an empty register set and is discarded at
    stream.cpp:315.

  - s_load_dwordx2 vcc, s[0:1], 0 uses the legal selector-encoded OPR_SREG destination, but
    to_register_ref() does not map its VCC selectors. classify_events() consequently treats the load
    as having no register result.

  Therefore these valid sequences can miss required scalar-memory waits:

  s_load_dword   ttmp0, s[0:1], 0
  s_mov_b32      s4, ttmp0

  s_load_dwordx2 vcc, s[0:1], 0
  s_cbranch_vccz 1

  The TTMP case can return a complete, empty report. The VCC case stops as incomplete at the branch
  but emits no hazard, despite stream.h (emulation/rocjitsu/lib/rocjitsu/src/rocjitsu/code/analysis/
  waitcheck/stream.h):64 promising that control-flow operands are checked. flat_scratch and xnack_mask
  selector destinations have the same representational problem.

  Suggested fix: provide waitcheck-specific tracking for TTMP lanes and selector-encoded special
  registers, or conservatively mark analysis incomplete whenever an operand cannot be represented. Add
  load-to-read and load-to-overwrite tests for TTMP and VCC, including a VCC control-flow consumer.

  Confidence: high. The instruction sequences and encodings were independently validated with LLVM’s
  gfx942 assembler.

**Review by atgutier — APPROVED** [review]
Approving with one more comment:

[MAJOR] emulation/rocjitsu/lib/rocjitsu/src/rocjitsu/code/analysis/waitcheck/stream.cpp:397 — Select
  the decoder by concrete target

  Reasoning: analyze_waitcheck_stream accepts an architecture but no target ID. For CDNA5,
  Decoder::create(arch) selects the default gfx1250 decoder. A valid gfx1251 stream containing
  v_pk_fma_f64 therefore fails decoding before the checker can analyze subsequent memory dependencies.

  Evidence: At the PR head, decoder.cpp:39–46 selects the default target for architecture lookup, and
  cdna5/target_descriptor.h:14–27,47 defines gfx1250 and gfx1251 with different instruction features.
  The existing cdna5_architecture_test.cpp:367–381 confirms that the architecture-default decoder
  rejects instructions accepted by the gfx1251 decoder.

  Suggested fix: Let callers provide the concrete target identity, use it to construct the decoder,
  and add a gfx1251 stream regression.

  Confidence: High. The existing decoder test establishes the failing input; the PR head was not run
  locally.
